tests/apps/test_embed.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Unit tests for strawberryfields.apps.clique
+"""
+# pylint: disable=no-self-use,unused-argument
+import numpy as np
+import pytest
+from strawberryfields.apps.train import embed
+
+feats = [
+    [[0.1, 0.2], [0.2, 0.1]],
+    [[0.1, 0.2, 0.3], [0.3, 0.1, 0.2], [0.2, 0.3, 0.1]],
+    [[0.1, 0.2, 0.3, 0.4], [0.4, 0.1, 0.2, 0.3], [0.3, 0.4, 0.1, 0.2], [0.2, 0.3, 0.4, 0.1]],
+]
+feats = np.array([np.array(f) for f in feats])
+
+ps = [[1.0, 2.0], [1.0, 2.0, 3.0], [1.0, 2.0, 3.0, 4.0]]
+ps = np.array([np.array(p) for p in ps])
+
+weights_f = np.array(
+    [
+        np.exp(-np.array([0.5, 0.4])),
+        np.exp(-np.array([1.4, 1.1, 1.1])),
+        np.exp(-np.array([3.0, 2.4, 2.2, 2.4])),
+    ]
+)
+weights = np.array(
+    [
+        np.exp(-np.array([1.0, 2.0])),
+        np.exp(-np.array([1.0, 2.0, 3.0])),
+        np.exp(-np.array([1.0, 2.0, 3.0, 4.0])),
+    ]
+)
+
+jacobian_f = np.array([np.zeros((d, d)) for d in range(2, 5)])
+jacobian = np.array([np.zeros((d, d)) for d in range(2, 5)])
+
+for i in range(3):
+    jacobian[i] = -np.diag(weights[i])
+    for j in range(i + 2):
+        for k in range(i + 2):
+            jacobian_f[i][j, k] = -feats[i][j, k] * weights_f[i][j]
+
+
+@pytest.mark.parametrize("dim", range(2, 5))
+class TestExpFeatures:
```

Comment:
Normally I'm used to splitting up the tests so that one function/class in the code corresponds to one test class, with methods of the test class being individual tests of the function/class in the code. Here we split up test classes based upon methods of the code class :thinking: I wonder if there is a "best practice" as I think the way you've done it works well.

Reply:
I don't think there is any particular best practice --- test organization can differ depending on the code being tested, sometimes it makes sense to organize it via functionality, sometimes via code unit, etc. In a lot of cases I've seen people recommend not using classes at all, and just to use test functions (as we do in the walrus).

Reply:
Thanks @josh146!